test/test_ufbrw.F90

```
@@ -0,0 +1,179 @@
+      enddo
+      call closbf(20)
+      write(55,*)
+      enddo
+
+! copy the testfile to test helpers getwin, invwin, newwin, nxtwin in output more
+! print the results in a text file for verification
+
+      open(20,file=file,form='unformatted')
+      open(50,file='ufbrw_bufr_out',form='unformatted')
+
+      call openbf(20,'IN ',20)
+      call openbf(50,'OUT',20)
+      
+      write(55,*);write(55,*)'read/write from unit 20'
+
+      do while(ireadmg(20,subset,idate)==0)
+        do while(ireadsb(20)==0)
+          call ufbint(20,arr,10,255,irt,'POB QOB TOB UOB VOB')
+          write(55,'(5(1x,f8.2))')arr(1:5,1:irt)
+          call openmb(50,subset,idate)
+          call ufbint(50,arr,10,irt,jrt,'POB QOB TOB UOB VOB')
+          call writsb(50)
+        enddo
+      enddo
+
+! read the testfile copy and print the results in a text file for verification
+
+      call closbf(50)
+      open(50,file='ufbrw_bufr_out',form='unformatted')
+      call openbf(50,'IN',50) 
+
+      write(55,*);write(55,*)'read/write from unit 50'
+
+      do while(ireadmg(50,subset,idate)==0)
+        do while(ireadsb(50)==0)
+          call ufbint(50,arr,10,255,irt,'POB QOB TOB UOB VOB')
+          write(55,'(5(1x,f8.2))')arr(1:5,1:irt)
+        enddo
+      enddo
+
+! close the testfile for writing and open it for reading
+
+      close(55)
+      open(55,file='ufbrw_prnt_out')
+
+! verify the testfile contents against output stored in brr array strings
+
+      do n=1,56     
+        read(55,'(a55)',iostat=iret) line
+        call strsuc(line  ,str1,len1)
+        call strsuc(brr(n),str2,len2)
+        if(n<=55.and.iret==0.and.str1/=str2) then
+          print*,n,len1,len2
+          print*,"str1:",str1  
+          print*,"str2:",str2 
+          call bort('error exit from ufbrw_test')
+        elseif(n>55.and.iret==0) then
+          print*,n,iret
+          call bort('error exit from ufbrw_test')
+        elseif(n==56) then
+          exit
+        endif
+      enddo
+
```

Thanks for this Jack.

The only comment I would add is that we've taken a different approach in all of our other test codes, which is to Fortran "stop" immediately with a unique non-zero number whenever there's an unexpected result, (and which in turn causes the test to automatically fail), rather than continuing to run through to the end of the code.

So I would recommend changing each of your `call bort('error exit from ufbrw_test')` lines to just be `stop n` and that way you're guaranteed a unique non-zero number which tells you immediately which line of brr didn't match its expected counterpart.  And then just change your `exit` line to `stop 57`.

And that way you also don't ever need to print the value of n either, because the stop number will already tell you that.

---

That works.